### PR TITLE
Implement identification algorithm

### PR DESCRIPTION
**Pull request by cthoyt**
This PR implements the ID and IDC algorithms. It introduces a few new data structures to hold the information that needs to get passed around a lot, as these are both recursive algorithms.

Several unused assets were removed in 5343a4c, in case you want to go back and check out the commented out old implementation from BEL2SCM and the unused test cases

### CODE REVIEW

File: tests/test_algorithm/test_id_alg.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# -*- coding: utf-8 -*-
```
Comment (cthoyt) [inline tests/test_algorithm/test_id_alg.py:1]: Don't forget there are example graphs available in https://github.com/y0-causal-inference/y0/blob/main/src/y0/examples.py

File: tests/test_algorithm/test_id_alg.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        )
+
+    def assert_identify(
+        self, expression: Expression, graph: NxMixedGraph, query: Expression
+    ):
+        """Assert that the graph returns the same."""
+        self.assert_expr_equal(expression, identify(graph, query))
+
+    def test_figure_2a(self):
+        """Test Figure 2A."""
```
Comment (cthoyt) [inline tests/test_algorithm/test_id_alg.py:42]: Would you please add the reference from which these figures' graphs are coming in the docstrings

Comment (djinnome) [reply tests/test_algorithm/test_id_alg.py]: Done

Comment (djinnome) [reply tests/test_algorithm/test_id_alg.py]: Shpitser, I., & Pearl, J. (2008). Complete Identification Methods for the Causal Hierarchy. Journal of Machine Learning Research.

File: tests/test_algorithm/test_id_alg.py (hunk 1 above)
Comment (djinnome) [reply tests/test_algorithm/test_id_alg.py:1]: I didn't know about this!  I will add the examples from   this paper:
```
      Shpitser, I., & Pearl, J. (2008). Complete Identification Methods for the Causal Hierarchy. Journal of Machine Learning Research. 
```

File: src/y0/examples.py [hunk 3]
```diff
@@ -136,6 +138,36 @@ class Example:
     graph=m_graph,
 )
 
+# Line 1 example
+line_1_example_a = Example(
+    name="Line 1 of ID algorithm",
+    reference="out of the mind of JZ",
```
Comment (cthoyt) [inline src/y0/examples.py:144]: Can we get a DOI for this? Would definitely come in handy ;)

File: src/y0/examples.py [hunk 4]
```diff
@@ -138,6 +142,167 @@ class Example:
...
+    graph=NxMixedGraph.from_edges(
+        directed=[
+            ("Z", "Y"),
+        ]
+    ),
+    identifications=[
+        dict(
+            id_in=[
```
Comment (cthoyt) [inline src/y0/examples.py:166]: is this dict always two keys? If that's the case, I'd suggest a named tuple data structure instead

File: src/y0/algorithm/identify/utils.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        cls,
+        query: Probability,
+    ) -> Query:
+        """Instantiate an identification.
+
+        :param query: The query probability expression
+        :returns: An identification tuple
+        """
+        outcomes, treatments = get_outcomes_and_treatments(query=query)
+        outcomes = {Variable(v.name) for v in outcomes}  # clean counterfactuals
```
Comment (cthoyt) [inline src/y0/algorithm/identify/utils.py:89]: @djinnome should outcomes be all of the parents + children of the distribution? or just the children

File: src/y0/algorithm/identify/utils.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        treatments = {Variable(v.name) for v in treatments}
+        conditions = {Variable(v.name) for v in query.distribution.parents}
+        return Query(
+            outcomes=outcomes,
+            treatments=treatments,
+            conditions=conditions,
+        )
+
+    def treat_condition(self, condition: Variable) -> Query:
```
Comment (cthoyt) [inline src/y0/algorithm/identify/utils.py:98]: @djinnome need a better name

File: src/y0/algorithm/identify/utils.py (hunk 5 above)
Comment (djinnome) [reply src/y0/algorithm/identify/utils.py:89]: `outcomes` should be children.   `conditions` should be parents

File: src/y0/algorithm/identify/utils.py (hunk 6 above)
Comment (djinnome) [reply src/y0/algorithm/identify/utils.py:98]: how about `exchange_observation_with_action`  which suggests an inverse operation `exchange_action_with_observation` as both are licensed by rule 2 of the do calculus

File: src/y0/algorithm/identify/utils.py (hunk 5 above)
Comment (cthoyt) [reply src/y0/algorithm/identify/utils.py]: Fixed in ce8a659

File: src/y0/algorithm/identify/utils.py (hunk 6 above)
Comment (cthoyt) [reply src/y0/algorithm/identify/utils.py]: Done in f682f44. I also implemented `exchange_action_with_observation`

File: src/y0/algorithm/identify/id_c.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        \newcommand\ci{\perp\!\!\!\perp}
+        \newcommand{\ubar}[1]{\underset{\bar{}}{#1}}
+        \newcommand{\obar}[1]{\overset{\bar{}}{#1}}
+        \text{if } (\exists Z \in \mathbf{Z})(\mathbf{Y} \ci Z | \mathbf{X}, \mathbf{Z}
+        - \{Z\})_{G_{\bar{\mathbf{X}}\ubar{Z}}} \\
+        \text{then } P(\mathbf{Y}|do(\mathbf{X}),\mathbf{Z}) = P(\mathbf Y|do(\mathbf X), do(Z), \mathbf{Z} - \{Z\})
+    """
+    graph = identification.graph
+    treatments = identification.treatments
+    conditions = treatments | (identification.conditions - {condition})
```
Comment (cthoyt) [inline src/y0/algorithm/identify/id_c.py:50]: does this operation `conditions = treatments | (identification.conditions - {condition})` have a name?

**Pull request merged**